### PR TITLE
NNI on Windows for NNI Remote mode

### PR DESCRIPTION
**Pull request by demianzhang**
#1053
Install nni on Windows, run experiments in remote linux machines
1. Use python tarfile to replace the tar command when compress the file
2. Change the windows path joined with '\\' into linux path joined with '/'
3. Fix nnictl stop using 'SIGBREAK' and remove the trialJobMetricListener to clean the experiments.

Pipeline:
copy nni project to remote and build wheel in remote, install the nni wheel in docker and finish the integration test.
1. change the pypi Makefile, using downloaded node and yarn to build wheel
2. use putty pscp command to get docker port

### CODE REVIEW

File: src/nni_manager/common/utils.ts [hunk 1]
```diff
@@ -459,6 +469,26 @@ function getNewLine(): string{
     }
 }
 
+/**
+ * Use '/' to join path instead of '\' for all kinds of platform
+ * @param path 
+ */
+function pathJoin(...paths: any[]): string{
```
Comment (QuanluZhang) [inline src/nni_manager/common/utils.ts]: mark

File: src/nni_manager/core/nnimanager.ts [hunk 2]
```diff
@@ -583,17 +584,19 @@ class NNIManager implements Manager {
...
         // TO DO: cannot run this method more than once in one NNIManager instance
         if (this.dispatcher === undefined) {
             throw new Error('Error: tuner or job maintainer have not been setup');
         }
-        this.trainingService.addTrialJobMetricListener((metric: TrialJobMetric) => {
-            this.onTrialJobMetrics(metric).catch((err: Error) => {
-                this.criticalError(new NNIError('Job metrics error', `Job metrics error: ${err.message}`, err));
-            });
-        });
+        this.trainingService.addTrialJobMetricListener(this.trialJobMetricListener);
```
Comment (QuanluZhang) [inline src/nni_manager/core/nnimanager.ts]: mark

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 3]
```diff
@@ -438,7 +437,7 @@ class RemoteMachineTrainingService implements TrainingService {
      */
     private getLocalGpuMetricCollectorDir(): string {
         let userName: string = path.basename(os.homedir()); //get current user name of os
-        return `${os.tmpdir()}/${userName}/nni/scripts/`;
+        return path.join(os.tmpdir(), userName, 'nni', 'scripts');
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: mark

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 4]
```diff
@@ -481,7 +480,7 @@ class RemoteMachineTrainingService implements TrainingService {
     private async initRemoteMachineOnConnected(rmMeta: RemoteMachineMeta, conn: Client): Promise<void> {
         // Create root working directory after ssh connection is ready
         await this.generateGpuMetricsCollectorScript(rmMeta.username); //generate gpu script in local machine first, will copy to remote machine later
-        const nniRootDir: string = `${os.tmpdir()}/nni`;
+        const nniRootDir: string = getRemoteTmpDir(this.remoteOS) + '/nni';
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: mark

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 5]
```diff
@@ -603,20 +602,20 @@ class RemoteMachineTrainingService implements TrainingService {
...
+            pathJoin(trialWorkingFolder, '.nni', 'code')
         )
 
         //create tmp trial working folder locally.
-        await cpp.exec(`mkdir -p ${path.join(trialLocalTempFolder, '.nni')}`);
+        await execMkdir(path.join(trialLocalTempFolder, '.nni'));
 
         //create tmp trial working folder locally.
-        await cpp.exec(`cp -r ${this.trialConfig.codeDir}/* ${trialLocalTempFolder}`);
+        await execCopydir(this.trialConfig.codeDir+'/*',trialLocalTempFolder);
```
Comment (QuanluZhang) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: mark

File: src/nni_manager/common/utils.ts [hunk 6]
```diff
@@ -459,6 +469,26 @@ function getNewLine(): string{
...
+    let dir:string = '';
+    for(let path of paths){
+        if(dir === ''){
+            dir = path;
+        }else{
+            if(dir === '/'){
+                dir = dir + path;
+            }else{
+                dir = dir + '/' + path;
```
Comment (QuanluZhang) [inline src/nni_manager/common/utils.ts]: potential bug

File: src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts [hunk 7]
```diff
@@ -575,18 +574,18 @@ class RemoteMachineTrainingService implements TrainingService {
...
 
         let command: string;
         // Set CUDA_VISIBLE_DEVICES environment variable based on cuda_visible_device
-        // If no valid cuda_visible_device is defined, set CUDA_VISIBLE_DEVICES to empty string to hide GPU device
+        // If no valid cuda_visible_device is defined, set CUDA_VISIBLE_DEVICES to '-1' to hide GPU device
         if(typeof cuda_visible_device === 'string' && cuda_visible_device.length > 0) {
             command = `CUDA_VISIBLE_DEVICES=${cuda_visible_device} ${this.trialConfig.command}`;
         } else {
-            command = `CUDA_VISIBLE_DEVICES=" " ${this.trialConfig.command}`;
+            command = `CUDA_VISIBLE_DEVICES='-1' ${this.trialConfig.command}`;
```
Comment (chicm-ms) [inline src/nni_manager/training_service/remote_machine/remoteMachineTrainingService.ts]: Suggest to keep the original CUDA_VISIBLE_DEVICES=" " which already tested.

File: test/pipelines-it-remote-windows.yml [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  - task: SSH@0
+    inputs:
+      sshEndpoint: remote_nni-ci-gpu-01
+      runOptions: inline
+      inline: cd /tmp/nnitest/$(Build.BuildId)/nni-remote/deployment/pypi;make build
+    continueOnError: true
+    displayName: 'build nni bdsit_wheel'
+  - task: SSH@0
+    inputs:
+      sshEndpoint: remote_nni-ci-gpu-01
```
Comment (chicm-ms) [inline test/pipelines-it-remote-windows.yml]: this endpoint remote_nni-ci-gpu-01  should not be hard coded.

File: test/pipelines-it-remote-windows.yml [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+      runOptions: commands
+      commands: python3 /tmp/nnitest/$(Build.BuildId)/nni-remote/test/remote_docker.py --mode start --name $(Build.BuildId) --image nni/nni --os windows
+    displayName: 'Start docker'
+  - task: DownloadSecureFile@1
+    inputs:
+      secureFile: remote_ci_private_key
```
Comment (chicm-ms) [inline test/pipelines-it-remote-windows.yml]: suggest to change this key file configurable from pipeline settings.

File: src/nni_manager/core/nnimanager.ts [hunk 10]
```diff
@@ -76,6 +77,11 @@ class NNIManager implements Manager {
             status: 'INITIALIZED',
             errors: []
         };
+        this.trialJobMetricListener = (metric: TrialJobMetric) => {
+            this.onTrialJobMetrics(metric).catch((err: Error) => {
+                this.criticalError(new NNIError('Job metrics error', `Job metrics error: ${err.message}`, err));
```
Comment (chicm-ms) [inline src/nni_manager/core/nnimanager.ts]: In order to handle string type err, please reference this PR to change it like this: ```this.criticalError(NNIError.FromError(err, 'Job metrics error: '));```

**Comment by QuanluZhang** [comment]
please update doc accordingly

File: test/remote_docker.py [hunk 11]
```diff
@@ -30,18 +30,29 @@ def find_wheel_package(dir):
...
+    if nnimanager_os == 'windows':
+        wheel_name = find_wheel_package(os.path.join(source_dir, 'nni-remote/deployment/pypi/dist'))
+    else:
+        wheel_name = find_wheel_package(os.path.join(source_dir, 'dist'))
+        
     if not wheel_name:
         print('Error: could not find wheel package in {0}'.format(source_dir))
         exit(1)
-    sdk_cmds = ['docker', 'exec', name, 'python3', '-m', 'pip', 'install', '/tmp/nni/dist/{0}'.format(wheel_name)]
+    def get_dist(wheel_name):
```
Comment (xuehui1991) [inline test/remote_docker.py]: add a blank line?

**Pull request merged**